### ISSUE TITLE
% is not considered to be a binary operator

### ISSUE THREAD

**Issue by FichteFoll** [PyCQA/pep8#422]
```py
print("%s"
      % "text")
```
does not emit W503.

The `%` operator is shown in the example for https://www.python.org/dev/peps/pep-0008/#multiline-if-statements, so it can be assumed that this is a recommendation, yet it is [excluded from the bin-op check](https://github.com/PyCQA/pep8/blob/151758ce8f29fd65e10dc0720446c5a3bb164602/pep8.py#L1026).

Even though I hate this convention and do not follow it, this is inconsistent.

**Comment by sigmavirus24** [comment]
No. We do not need a separate error code.

**Comment by sigmavirus24** [comment]
Allow me to elaborate: There's a comment there explaining the rationale for why `%` is excluded. I don't agree with the rationale, but if we're going to fix this (which I agree, we should) then we should do this in pep8 1.7.0

**Comment by jayvdb** [comment]
fwiw, the use of % in a string context was explicitly added as an example. https://hg.python.org/peps/rev/973665d69686

And it appears the cython code base is mostly using the trailing % style:
```
cpython$ find . -name '*.py' | xargs grep '^ *% ' | wc -l
257
cpython$ find . -name '*.py' | xargs grep '["'\''] %$' | wc -l
429
```

However `Lib/asyncio` is newish code which mostly places the % on the new line.

IMO there should be two codes; one for maths binary operators (inclusive of %), and a new code for string binary operators (i.e. just %) which is deprecated by PEP 3101.

**Comment by sigmavirus24** [comment]
> And it appears the cython code base is mostly using the trailing % style:

It's repeatedly said that the CPython source is not (nor will it likely ever be) fully pep8 compliant. I don't see a point in using that as a baseline for this decision.

> IMO there should be two codes; one for maths binary operators (inclusive of %), and a new code for string binary operators (i.e. just %) which is deprecated by PEP 3101.

So I take it you're up to do the work then?

**Comment by jayvdb** [comment]
@sigmavirus24, sure I'll do it, if you're agreeable with my suggestion regarding keeping the two cases separate.

**Comment by xZise** [comment]
Unless PEP8 actually executes code there are cases where it's not possible to separate between the modulo operation or string format operation:

  def percent(a, b):
      return a % b

    print(percent(42, 10))
    print(percent('%s', 10))

Now an %-operation like here can't be determined if it's modulo or string format. And in fact most modulo operations are probably like `a % 2` and that could also be a string operation when `a = '%s'`. So you could either say that if it can't determine what type of operation it is, it assumes it's modulo. But that behavior could probably confuse others:

    msg = '%s'
    # fine
    print(msg %
          42)
    # not fine
    print(msg
          % 42)
    # but this is fine, for someone who doesn't know the rules might wonder
    # why the above is then not fine
    print('%s'
          % 42)

And if you assume it's a string format operation, it'll probably assume in most cases this even though it's a modulo operation. While it's probably not as confusing it's not really useful if in most cases it's ignored anyway. Afaik the only sensible solution is to add an option to ignore the %-operator from the binary operator cases.

Also I don't agree that having the operator in string format in the next line is compliant to the PEP8, which is what this bug report is about. I won't stop you from implementing a logic that tries to detect if it's string format, but if it defaults to modulo operator if it can't detect it, it's afaik only a step between ignoring the %-operator at all (like now) or always applying the binary operator rule to it (which is suggested in the opening post).

This by the way sounds similar to GreenSteam/pep257#96 and GreenSteam/pep257#91 as both have multiple interpretations to a rule. Here it's “is % a binary operator” and on pep257 the number of lines above a class docstring.

**Comment by sigmavirus24** [comment]
> @sigmavirus24, sure I'll do it, if you're agreeable with my suggestion regarding keeping the two cases separate.

@jayvdb If you can make pep8 so accurate as to not generate false positives that will drive people to stop using pep8, I'm happy to not be as vocally opposed to two error codes. That said, @IanLee1521 is the maintainer, not me.

The fact is that what @xZise says next is very accurate and kind of my reasoning for why two error codes will be confusing, error prone, and a poor choice:

> Unless PEP8 actually executes code there are cases where it's not possible to separate between the modulo operation or string format operation

The thing is that pep8 the tool has always stuck to being as fast as possible. This means tokenizing the source given to it for analysis, no parsing into an AST, and no execution of arbitrary code. It's simplest to just roll this into the existing error code. I don't see a point in having a separate error code. What will happen is that people have ignored the existing error code will have a new one pop up and will have to ignore that too. This will be far more disruptive than you think. It'll cause people to wonder why there are two separate error codes for what is effectively the same check. It's not worth the confusion at all to add a second check.

**Comment by FichteFoll** [comment]
Will this issue be relevant for #502?

**Comment by FichteFoll** [comment]
PEP 8 was in favor of the opposite for the longest time, which is probably why W504 wasn't made default.

**Comment by sigmavirus24** [comment]
> PEP 8 was in favor of the opposite for the longest time, which is probably why W504 wasn't made default.

exactly

**Comment by mats-erik** [comment]
> 
> 
> PEP 8 was in favor of the opposite for the longest time, which is probably why W504 wasn't made default.

I'm aware of that. The style guide is improved as time goes and I think pycodestyle default ignore setting should be updated to stay up-to-date with the style guide, at least when is comes to rules that have examples with Yes/No.

**Comment by IanLee1521** [comment]
I think it's reasonable in the next major version to switch the default behave. I'll leave this issue open and mark it as for v3.x

**Comment by sigmavirus24** [comment]
> I'm aware of that. The style guide is improved as time goes and I think pycodestyle default ignore setting should be updated to stay up-to-date with the style guide, at least when is comes to rules that have examples with Yes/No.

No, that would be ridiculously bad for users. Consider someone expecting W503 to be enforced. If we in a minor version (e.g., 2.1.1 -> 2.2.0) say W503 is off-by-default and surprise you have W504 that you need to update your entire code-base to follow, that's crap. That breaks CI, wastes developer time, and will cause people to stop using the tool in favor of something else. I don't work on this tool to make people's lives worse.

**Comment by mats-erik** [comment]
> 
> 
> > I'm aware of that. The style guide is improved as time goes and I think pycodestyle default ignore setting should be updated to stay up-to-date with the style guide, at least when is comes to rules that have examples with Yes/No.
> 
> No, that would be ridiculously bad for users. Consider someone expecting W503 to be enforced. If we in a minor version (e.g., 2.1.1 -> 2.2.0) say W503 is off-by-default and surprise you have W504 that you need to update your entire code-base to follow, that's crap. That breaks CI, wastes developer time, and will cause people to stop using the tool in favor of something else. I don't work on this tool to make people's lives worse.

I would not expect/like such changes in minor/micro releases either, but IanLee1521 added this to the 3.0 milestone which I think is great. 
If pycodestyle defaults are kept up-to-date with the style guide, then when using pycodestyle on new codebases one would not need to override the defaults make pycodestyle help with keeping the codebase follow the current pep 8 recommendations. In existing codebases one may add a config in the project root to override the defaults if there are too many W504 warnings to clean up.
Since both W503 and W504 are ignored by default right now because pep 8 was not clear on this earlier, someone expecting W503 to be enforced is already using a config to remove W503 from the ignore-list, which means removing W504 from the default ignore won't change anything for him/her as the ignore setting is already overridden.
